pylops/signalprocessing/DWT.py

```diff
@@ -0,0 +1,144 @@
+import logging
+from math import log, ceil
+
+import numpy as np
+from pylops import LinearOperator
+from pylops.basicoperators import Pad
+
+try:
+    import pywt
+except ModuleNotFoundError:
+    pywt = None
+
+logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.WARNING)
+
+
+def _checkwavelet(wavelet):
+    """Check that wavelet belongs to pywt.wavelist
+    """
+    wavelist = pywt.wavelist(kind='discrete')
+    if wavelet not in wavelist:
+        raise ValueError("'%s' not in family set = %s" % (wavelet,
+                                                          wavelist))
+
+def _adjointwavelet(wavelet):
+    """Define adjoint wavelet
+    """
+    waveletadj = wavelet
+    if 'rbio' in wavelet:
+        waveletadj = 'bior' + wavelet[-3:]
+    elif 'bior' in wavelet:
+        waveletadj = 'rbio' + wavelet[-3:]
+    return waveletadj
+
+
+class DWT(LinearOperator):
+    """One dimensional Wavelet operator.
+
+    Apply 1D-Wavelet Transform along a specific direction ``dir`` of a
+    multi-dimensional array of size ``dims``.
+
+    Note that the Wavelet operator is an overload of the ``pywt``
+    implementation of the wavelet transform. Refer to
+    https://pywavelets.readthedocs.io for a detailed description of the
+    input parameters.
+
+    Parameters
+    ----------
+    dims : :obj:`int` or :obj:`tuple`
+        Number of samples for each dimension
+    dir : :obj:`int`, optional
+        Direction along which FFT is applied.
+    wavelet : :obj:`str`, optional
+        Name of wavelet type. Use :func:`pywt.wavelist(kind='discrete')` for
+        a list of
+        available wavelets.
+    level : :obj:`int`, optional
+        Number of scaling levels (must be >=0).
+    dtype : :obj:`str`, optional
+        Type of elements in input array.
+
+    Attributes
+    ----------
+    shape : :obj:`tuple`
+        Operator shape
+    explicit : :obj:`bool`
+        Operator contains a matrix that can be solved explicitly
+        (True) or not (False)
+
+    Raises
+    ------
+    ModuleNotFoundError
+        If ``pywt`` is not installed
+    ValueError
+        If ``wavelet`` does not belong to ``pywt.families``
+
+    Notes
+    -----
+    The Wavelet operator applies the multilevel Discrete Wavelet Transform
+    (DWT) in forward mode and the multilevel Inverse Discrete Wavelet Transform
+    (IDWT) in adjoint mode.
+
+    Wavelet transforms can be used to compress signals and present
+    a key advantage over Fourier transforms in that they captures both
+    frequency and location information in time. Consider using this operator
+    as sparsifying transform when using L1 solvers.
+
+    """
+    def __init__(self, dims, dir=0, wavelet='haar', level=1, dtype='float64'):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'dir'](https://app.codacy.com/app/mrava87/pylops/pullRequest?prid=4678589)